packages/dds/tree/src/changeset/visit.ts

```diff
@@ -0,0 +1,209 @@
+ * Note that while the attach positions are expressed top-down, there is still a bottom-up spirit to building trees
+ * that are being inserted.
+ *
+ * The second challenge, is that of the inability of the visitor to move-in content that has yet to be moved-out.
+ * This leads to a two-pass algorithm, but there are two degrees for freedom to consider:
+ * 1) Whether inserts should be performed in the first pass whenever possible (some are not: inserts below a move-ins
+ *   for which we have not yet seen the matching move-out).
+ * Pros: The path above the insertion point is walked once instead of twice
+ * Cons: The paths within the inserted content risk being walked twice instead of once (once for building the content,
+ * once for traversing the tree to reach move-in marks in the second phase).
+ *
+ * 2) Whether move-ins for which we have the move-out content should be performed in the first pass.
+ * Pros: The path above the move-in point is walked once instead of twice
+ * Cons: We now have to record which of the move-ins we did not perform in the first pass. We could build a trie of
+ * those to reduce the amount of sifting we have to do on the second pass.
+ *
+ * The presence of a move table, which lists the src and dst paths for each move, could be leveraged to make some of
+ * these option more efficient:
+ * - If inserts are allowed in the first pass and move-ins are not allowed in the first pass, then the move table
+ *   describes exactly which parts of the delta need applying in the second pass.
+ * - If inserts and move-ins are allowed in the first pass then having a boolean flag for each entry in the move table
+ *   that describes whether the move has been attached, or having a set for that describes which entries remain, would
+ *   describe which parts of the delta  need applying in the second pass.
+ *
+ * Current implementation:
+ * - Performs inserts in the first pass
+ * - Does not perform move-ins in the first pass
+ * - Skips the second pass if no move-outs were encountered in the first pass
+ * - Does not leverage the move table
+ *
+ * Future work:
+ * - Take a path to the subtree of interest (or some other description of the regions of interest)
+ * - Leverage move table when it gets added to Delta
+*/
+
+/**
+ * Crawls the given `delta`, calling `visitor`'s callback for each change encountered.
+ * Each successive call to the visitor callbacks assumes that the change described by earlier calls have been applied
+ * to the document tree. For example, for a change that deletes the first and third node of a field, the visitor calls
+ * will pass indices 0 and 1 respectively.
+ * @param delta - The delta to be crawled.
+ * @param visitor - The object to notify of the changes encountered.
+ */
+export function visitDelta(delta: Delta.Root, visitor: DeltaVisitor): void {
+	const moveInfo: MoveOutInfo = new Map();
+	firstPass(delta, { visitor, moveInfo });
```

Comment:
this repo generally uses space indent (though we have no linter enforcement)

Reply:
Thanks for bringing this up, I had no idea. In fact I'm quite surprised because I thought MS has unversally adopted tabs for the sake of accessibility. I found an old issue (#313) that seems to imply we have/had ESlint rules for tabs/spaces. I'll try to find out more.

Reply:
Looks like the only rule we have disables ESLint checking of tabs vs spaces. I've updated the files in this PR to use spaces.